source/material_model/rheology/elasticity.cc

```diff
@@ -269,80 +355,199 @@ namespace aspect
-                stress_new = ( ( 1. - ( dt / dte ) ) * stress_old ) + ( ( dt / dte ) * stress_new ) ;
+      // Fill the function that computes the reaction rates for the operator
+      // splitting step that at the beginning of the new timestep updates the
+      // stored compositions $tau^{0\mathrm{adv}}$ at time $t$ to $tau^{t}$.
+      template <int dim>
+      void
+      Elasticity<dim>::fill_reaction_rates (const MaterialModel::MaterialModelInputs<dim> &in,
+                                            const std::vector<double> &average_elastic_shear_moduli,
+                                            MaterialModel::MaterialModelOutputs<dim> &out) const
+      {
+        ReactionRateOutputs<dim> *reaction_rate_out = out.template get_additional_output<ReactionRateOutputs<dim>>();
 
-                // Fill reaction terms
-                for (unsigned int j = 0; j < SymmetricTensor<2,dim>::n_independent_components ; ++j)
-                  out.reaction_terms[i][j] = -stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)]
-                                             + stress_new[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)];
+        if (reaction_rate_out == nullptr ||
+            (!this->get_parameters().use_operator_splitting && !((this->get_parameters().mapped_particle_properties).count(this->introspection().compositional_index_for_name("ve_stress_xx")))))
+          return;
 
+        // At the moment when the reaction rates are required (at the beginning of the timestep),
+        // 'solution' holds the solution of the previous timestep ($\tau^{t}_{0adv}$), and is the same as 'old_solution'.
+        // MaterialModelInputs are based on 'solution' when calling the MaterialModel for the reaction rates.
+        // This means that we can use 'in' to get to the $\tau^{t}_{0adv}$ and velocity/strain rate of the
+        // previous timestep.
+        if (in.current_cell.state() == IteratorState::valid && this->get_timestep_number() > 0)
+          {
+            for (unsigned int i = 0; i < in.n_evaluation_points(); ++i)
+              {
+                // Set all reaction rates to zero
+                for (unsigned int c = 0; c < in.composition[i].size(); ++c)
+                  reaction_rate_out->reaction_rates[i][c] = 0.0;
+
+                // Get $\tau^{0adv}$ of the previous timestep t from the compositional fields.
+                // This stress includes the rotation and advection.
+                SymmetricTensor<2, dim> stress_0_t;
+                for (unsigned int j = 0; j < SymmetricTensor<2, dim>::n_independent_components; ++j)
+                  stress_0_t[SymmetricTensor<2, dim>::unrolled_to_component_indices(j)] = in.composition[i][j];
+
+                // Get the old stress that is used to interpolate to timestep $t+\Delta t_c$. It is stored on the
+                // second set of n_independent_components fields, e.g. in 2D on field 3, 4 and 5.
```

Comment:
This is not entirely true. The old_stress in in.composition was advected but not rotated into the previous timestep. Below it will be updated to the full stress of the previous timestep, so that it can be advected into the current timestep (but again not rotated).

Reply:
Ok good to know. Could you update the comment to say that?